### PR TITLE
test: topology: verify that group 0 and token ring are consistent

### PR DESCRIPTION
**Pull request by kbr-scylla**
After topology changes like removing a node, verify that the set of group 0 members and token ring members is the same.

Fixes: #12153

### CODE REVIEW

**Review by kostja — APPROVED** [review]
I don't think you need to add this verification to every trivial test, perhaps it's enough to add to a complicated test. Please consider this, approving regardless of your decision.

**Comment by kbr-scylla** [comment]
> I don't think you need to add this verification to every trivial test, perhaps it's enough to add to a complicated test. Please consider this, approving regardless of your decision.

I agree - no point adding it everywhere - but I think it's good to have it in each of these existing tests, which cover the basic scenarios such as decommission, removenode, bootstrap, replace. It's fine to not add it to new test cases.

**Comment by kbr-scylla** [comment]
In fact the check appears to be failing in the replace tests, at least on CI... Interesting.

**Comment by kbr-scylla** [comment]
v2:
- after bootstrap/replace, the token ring is not guaranteed to be updated immediately (the tokens of the new node in NORMAL state take a few seconds to propagate through gossip) - take that into account and wait until token ring members match group 0 members.
- in `get_token_ring_host_ids`, get only the NORMAL members by using the `/storage_service/tokens_endpoint` REST API.

**Comment by kbr-scylla** [comment]
@xemul I don't typically request you to review my test patches, but since you started asking good questions about group 0 and how it integrates with the rest of the system - perhaps reviewing this patch would be of an interest to you?

File: test/topology/test_topology.py [hunk 1]
```diff
@@ -117,41 +160,46 @@ async def no_joining_nodes():
...
 
 @pytest.mark.asyncio
 async def test_replace_different_ip(manager: ManagerClient, random_tables) -> None:
     servers = await manager.running_servers()
     await manager.server_stop(servers[0].server_id)
     replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = False, use_host_id = False)
     await manager.server_add(replace_cfg)
-    # TODO: check that group 0 no longer contains the replaced node (#12153)
+    servers = await manager.running_servers()
+    await wait_for_token_ring_and_group0_consistency(manager, servers, time.time() + 30)
```
Comment (xemul) [inline test/topology/test_topology.py:175]: nit: all invocations of `wait_for_..._consistency` are preceded by the `await manager.running_servers()` call, so it's probably worth moving the latter call into the former

**Comment by kbr-scylla** [comment]
v2:
- `check_token_ring_and_group0_consistency` and `wait_for_token_ring_and_group0_consistency` no longer take a `servers` parrameter, they take running servers from `manager`

File: test/topology/test_topology.py (hunk 1 above)
Comment (kbr-scylla) [reply test/topology/test_topology.py]: Done

File: test/topology/test_topology.py [hunk 2]
```diff
@@ -117,41 +161,46 @@ async def no_joining_nodes():
...
-    # TODO: check that group 0 no longer contains the decommissioned node (#12153)
 
 @pytest.mark.asyncio
 async def test_replace_different_ip(manager: ManagerClient, random_tables) -> None:
     servers = await manager.running_servers()
     await manager.server_stop(servers[0].server_id)
     replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = False, use_host_id = False)
     await manager.server_add(replace_cfg)
-    # TODO: check that group 0 no longer contains the replaced node (#12153)
+    servers = await manager.running_servers()
```
Comment (xemul) [inline test/topology/test_topology.py:175]: `servers` is unused

File: test/topology/test_topology.py [hunk 3]
```diff
@@ -117,41 +161,46 @@ async def no_joining_nodes():
...
+    await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
 
 @pytest.mark.asyncio
 async def test_replace_different_ip_using_host_id(manager: ManagerClient, random_tables) -> None:
     servers = await manager.running_servers()
     await manager.server_stop(servers[0].server_id)
     replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = False, use_host_id = True)
     await manager.server_add(replace_cfg)
-    # TODO: check that group 0 no longer contains the replaced node (#12153)
+    servers = await manager.running_servers()
```
Comment (xemul) [inline test/topology/test_topology.py:184]: ditto

File: test/topology/test_topology.py [hunk 4]
```diff
@@ -117,41 +161,46 @@ async def no_joining_nodes():
...
+    await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
 
 @pytest.mark.asyncio
 async def test_replace_reuse_ip(manager: ManagerClient, random_tables) -> None:
     servers = await manager.running_servers()
     await manager.server_stop(servers[0].server_id)
     replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = True, use_host_id = False)
     await manager.server_add(replace_cfg)
-    # TODO: check that group 0 no longer contains the replaced node (#12153)
+    servers = await manager.running_servers()
```
Comment (xemul) [inline test/topology/test_topology.py:193]: ditto

File: test/topology/test_topology.py [hunk 5]
```diff
@@ -117,41 +161,46 @@ async def no_joining_nodes():
...
+    await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
 
 @pytest.mark.asyncio
 async def test_replace_reuse_ip_using_host_id(manager: ManagerClient, random_tables) -> None:
     servers = await manager.running_servers()
     await manager.server_stop(servers[0].server_id)
     replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = True, use_host_id = True)
     await manager.server_add(replace_cfg)
-    # TODO: check that group 0 no longer contains the replaced node (#12153)
+    servers = await manager.running_servers()
```
Comment (xemul) [inline test/topology/test_topology.py:202]: ditto

File: test/topology/test_topology.py [hunk 6]
```diff
@@ -190,6 +239,9 @@ async def test_nodes_with_different_smp(manager: ManagerClient, random_tables: R
     logger.info(f'Adding --smp=5 server')
     await manager.server_add(cmdline=['--smp', '5'])
 
+    servers = await manager.running_servers()
```
Comment (xemul) [inline test/topology/test_topology.py:242]: ditto

**Comment by kbr-scylla** [comment]
Fixed, diff: https://github.com/scylladb/scylladb/compare/9d29bf837ee5c98785d1d371ca9448bc17f26b7e..a3a5f981edb174d5e964017735c7772b496b9c92

**Pull request closed without merging**